src/osc/query/oscquerydescription.cpp

```diff
@@ -0,0 +1,190 @@
+#include "osc/query/oscquerydescription.h"
+
+#include <QDebug>
+#include <QFile>
+#include <QJsonDocument>
+#include <QJsonValue>
+
+#include "preferences/configobject.h"
+
+namespace {
+
+template<typename T>
+T take_back(std::vector<T>* pVec) {
+    T last_element = std::move(pVec->back());
+    pVec->pop_back();
+    return last_element;
+}
+
+QString toOscAddress(const ConfigKey& key) {
+    QString oscAddress = key.group + key.item;
+    oscAddress.remove(QChar('['));
+    oscAddress.remove(QChar(']'));
+    oscAddress.replace(QChar('_'), QChar('/'));
+    return oscAddress;
+}
+
+} // namespace
+
+OscQueryDescription::OscQueryDescription()
+        : m_rootObject({{"DESCRIPTION", "Mixxx OSC root node"},
+                  {"FULL_PATH", "/"},
+                  {"ACCESS", "0"},
+                  {"CONTENTS", QJsonObject()}}) {
+}
+
+void OscQueryDescription::addAddress(
+        const QString& address,
+        const QString& type,
+        const QString& access,
+        const QString& description) {
+    QStringList pathParts = address.split('/');
+    QString fullPath;
+
+    std::vector<QJsonObject> objects;
+    objects.reserve(pathParts.size());
+
+    // Create a list of QJsonObjects.
+    // Note: QJsonObject is only temporary helper class for writing a QJsonValue
+    for (const auto& pathPart : std::as_const(pathParts)) {
+        if (pathPart.isEmpty()) {
+            continue;
+        }
+        fullPath += "/" + pathPart;
+        if (objects.size()) {
```

Comment:
implicit narrowing cast. what does this imply? this code is only executed on first iteration?

Reply:
Yes.